lib/broadway/producer.ex

```diff
@@ -171,14 +187,63 @@ defmodule Broadway.Producer do
   end
 
   @impl true
+  def handle_info({__MODULE__, :cancel_consumers}, %{rate_limiting: %{} = rate_limiting} = state) do
+    rate_limiting =
+      if :queue.is_empty(rate_limiting.message_buffer) do
+        cancel_consumers(state)
+        rate_limiting
+      else
+        %{rate_limiting | draining?: true}
+      end
+
+    {:noreply, [], %{state | rate_limiting: rate_limiting}}
+  end
+
   def handle_info({__MODULE__, :cancel_consumers}, state) do
-    for from <- state.consumers do
-      send(self(), {:"$gen_producer", from, {:cancel, :shutdown}})
+    cancel_consumers(state)
+    {:noreply, [], state}
+  end
+
+  def handle_info({__MODULE__, :handle_next_demand}, state) do
```

Comment:
So I just realized that when we arrive on this clause, the rate_limiting can be already closed because of another handle_demand. But it is fine, handle_demand will proceed to close it immediately. I wouldn't change anything for now, just a note.